lib/elixir/lib/enum.ex

```diff
@@ -789,11 +790,19 @@ defmodule Enum do
       iex> Enum.flat_map([{1,3}, {4,6}], fn({x,y}) -> x..y end)
       [1, 2, 3, 4, 5, 6]
 
+      iex> Enum.flat_map([1,2,3], fn(x) ->
+      ...>    if x > 2, do: :halt, else: [x, x]
+      ...> end)
+      [1, 1, 2, 2]
+
   """
   @spec flat_map(t, (element -> t)) :: list
   def flat_map(collection, fun) do
     reduce(collection, [], fn(entry, acc) ->
-      reduce(fun.(entry), acc, &[&1|&2])
+      case fun.(entry) do
```

Comment:
This code is not correct. Here is an example that will fail:

```
Enum.flat_map([1, 2, 3, 2], fn(x) ->
   if x > 2, do: :halt, else: [x, x]
end)
#=> [1, 1, 2, 2, 2, 2]
```

It should return only `[1, 1, 2, 2]`. To implement this, we would actually need to call `Enumerable.reduce/3`.